### PR TITLE
Add spells for Punk bids

### PR DESCRIPTION
**Pull request by catherine-2**
Brief comments on the purpose of your changes:

**For Dune Engine V2**

I've checked that:

### General checks:
* [ ] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [ ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [ ] the filename is unique and ends with .sql
* [ ] each sql file is a select statement and has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Comment by catherine-2** [comment]
@jeff-dude  I'm working on one more spell for CryptoPunks bids! 

The main test is passing and should be good based on the queries I wrote first but the hooks test is failing. It looks like this could possibly be related to the PR below that was merged earlier today? I didn't change anything in the yml file besides adding the new file info and it looks like other recently opened PRs have the same test failure issue.

Do you know of a fix to get around this? Would like to double check the test schema tables in Dune before requesting review / merge but they aren't loading due to the hook test failure.

Thank you!

https://github.com/duneanalytics/spellbook/pull/2234

**Comment by Hosuke** [comment]
> The main test is passing and should be good based on the queries I wrote first but the hooks test is failing. It looks like this could possibly be related to the PR below that was merged earlier today? I didn't change anything in the yml file besides adding the new file info and it looks like other recently opened PRs have the same test failure issue.
> 
> Do you know of a fix to get around this? Would like to double check the test schema tables in Dune before requesting review / merge but they aren't loading due to the hook test failure.

Possible cause:
env variable FILES with directory starts with `models/...` but not `spellbook/models/...`
Please see [the test log here](https://github.com/duneanalytics/spellbook/actions/runs/3636085842/jobs/6135751117#step:8:59).

We may fix this by adding or removing `spellbook/` prefix in the checking script.

**Comment by catherine-2** [comment]
> > The main test is passing and should be good based on the queries I wrote first but the hooks test is failing. It looks like this could possibly be related to the PR below that was merged earlier today? I didn't change anything in the yml file besides adding the new file info and it looks like other recently opened PRs have the same test failure issue.
> > Do you know of a fix to get around this? Would like to double check the test schema tables in Dune before requesting review / merge but they aren't loading due to the hook test failure.
> 
> Possible cause: env variable FILES with directory starts with `models/...` but not `spellbook/models/...` Please see [the test log here](https://github.com/duneanalytics/spellbook/actions/runs/3636085842/jobs/6135751117#step:8:59).
> 
> We may fix this by adding or removing `spellbook/` prefix in the checking script.

Ah ok got it - thank you! I've been working from my repository which is in a folder called abstractions still from when I originally forked. 

Is this the best course of action? I'm not sure how to edit my current file / branch without breaking things.
- Delete my repo
- Re-fork on the spellbook version
- Set up git one level higher (or should it be in the spellbook directory? wasn't sure with the spellbook/ requirement )

**Comment by catherine-2** [comment]
Or if it is possible to merge with that test failure and I can do that file reset I outlined before my next PR that would be great!

I just checked the test schema files and this PR looks good to merge from my perspective otherwise.

Thank you!

**Comment by Hosuke** [comment]
Hi @catherine-2 , you can ignore the pre-commit hook test result for now, as the PR passes dbt-test(which shows the code/spell can be compiled and built to models). I think the code in the issue may link to script outside the spellbook project. If that is a general issue, we will fix that in the main branch very soon.

Meanwhile we can still use the dbt-test result to check the spell.

**Comment by catherine-2** [comment]
@Hosuke @jeff-dude  This is ready for review / merge!  Details on the changes:

- 2 new incremental models for bid events + offer events. These are not exposed but power other views that are exposed
- 1 new view for "current_bids" that shows all bids that are currently open. There isn't event logging for auto cancelled bids so the logic mapped out follows the contract logic for when bids are cancelled. I confirmed the data output is accurate.
- 1 new view for "all_events" that combines trades, transfers, offers, bids and sales. This is exposed so wizards can query all of the CryptoPunks activity in one place

I have checked all of the tables on Dune using the test schema.

Please let me know of any questions / comments! I believe this should be the last PR for CryptoPunks (unless any bugs come up). Planning on sharing publicly after this is merged so other wizards can leverage these tables!

**Comment by catherine-2** [comment]
Read through everything one last time and made a couple of quick edits to formatting to match what is on cryptopunks.app. Done editing now!

**Comment by catherine-2** [comment]
Note: Please ignore for tonight - going to make a bunch of changes now that all of the CryptoPunks spells are complete to better optimize the code before we finalize!

**Comment by catherine-2** [comment]
@jeff-dude  This is now officially ready for review :) 

These are the final CryptoPunks spells so I went through and optimized other models as well. As a result the PR became much larger than anticipated (sorry!) so I wrote up an expanded description for all CryptoPunks files + changes in this PR. Please let me know of any questions or feedback - thank you!

**Incremental Models**

1. events

- This model is built to power cryptopunks_ethereum_trades, nft_events and nft_trades 
- In this PR I made a small edit to add evt_index

2. punk_bid_events (new in this PR)

- New incremental model for all bids entered and withdrawn on the CryptoPunks marketplace. Think this one is pretty straightforward

3. punk_offer_events (new in this PR)

- New incremental model for all offers entered and withdrawn on the CryptoPunks marketplace
- One question for you on this one: the 'PunkNoLongerForSale' event is logged when someone (a) cancels an offer, (b) transfers a punk or (c) sells a punk. I only want to capture (a) here - direct cancellations. I added logic "where evt_tx_hash not in ([pull from trades view and pull from punk_transfers model])". During the tests I noticed this model is running really slowly as a result. Wanted to see if that will cause any problems and/or there is a better way to structure this so it builds more quickly?

4. punk_transfers

- This model pulls all punk transfers
- Made a quick update to improve this model since I realized I should use the Assign call data directly vs. the static CTE I had copied in from this data previously 

**Views**

1. all_events (new in this PR)

- Combines bid, offer, trade and transfer data in one table so wizards can query all CryptoPunks events in one place. Structured to match the formatting / history as displayed on cryptopunks.app (the CryptoPunks marketplace website)
- One note on this one: I added in a table for Rarible sales directly. I spoke with Hildobby and he is already working on finalizing the Rarible spell for nft_trades. There were ~1,500 Rarible trades for CryptoPunks in late 2020 so it is important the data is in here. I know this structure is not ideal but would be great to keep + I will remove once we have the Rarible spell merged

2. current_bids (new in this PR)

- New view that looks at what bids are currently live
- There is no event logging for bids that are automatically cancelled by the contract (only manual cancellations by the user who created a bid are logged). As a result I built logic that mirrors the logic in the CryptoPunks contract to find what bids have been cancelled. I confirmed that this query is showing accurate data that matches what is shown in the marketplace

3. current_listings

- View that shows currently available listings
- Made minor updates in this PR to syntax (more explicit casting + slight changes to column names) and use the incremental models we built vs. the decoded contracts previously. No change to the logic

4.. current_owners_per_token

- Current owners for each token. No changes made in this PR

5. floor_price_over_time

- Floor price over time in both ETH and $. Made the same changes as current_listings above to syntax + use the incremental models in this PR

6. listings_over_time

- Listings over time. Made the same changes as current_listings & floor_price_over_time above to syntax + use the incremental models in this PR

7. owners_over_time

- Unique owners each day over time. No changes made in this PR

8. trades

- View on top of events that powers nft_trades. Only change in this PR was to add evt_index

**Static Table**

1. metadata

- Traits for each CryptoPunk. No changes made in this PR

**Comment by catherine-2** [comment]
@jeff-dude wanted to bump this for review whenever you have a chance! 

I believe it should be ready to merge. I tested the schema tables for every table/view I edited to confirm the data looks as expected.

File: models/cryptopunks/ethereum/cryptopunks_ethereum_current_listings.sql [hunk 1]
```diff
@@ -8,74 +8,65 @@
...
             , evt_block_number
             , evt_index
             , evt_block_time
             , evt_tx_hash
-    from {{ source('cryptopunks_ethereum','CryptoPunksMarket_evt_PunkTransfer') }}
+    from {{ ref('cryptopunks_ethereum_punk_transfers') }}
+)
+, latest_eth_price as (
+    select price
+    from prices.usd
```
Comment (jeff-dude) [inline models/cryptopunks/ethereum/cryptopunks_ethereum_current_listings.sql:51]: can you add the jinja syntax here?

File: models/cryptopunks/ethereum/cryptopunks_ethereum_events.sql [hunk 2]
```diff
@@ -153,8 +150,8 @@ inner join {{ source('ethereum','transactions') }} tx on a.evt_tx_hash = tx.hash
 {% endif %}
 
 left join {{ source('prices', 'usd') }} p on p.minute = date_trunc('minute', a.evt_block_time)
-    AND p.contract_address = "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"
-    AND p.blockchain = "ethereum"
+    and p.contract_address = "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"
```
Comment (jeff-dude) [inline models/cryptopunks/ethereum/cryptopunks_ethereum_events.sql:153]: in other scripts, you changed these to single quotes -- can you look to make consistent?

File: models/cryptopunks/ethereum/cryptopunks_ethereum_punk_offer_events.sql [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            , case when toAddress = '0x0000000000000000000000000000000000000000' then cast(NULL as varchar(5))
+                else toAddress end as to
+            , minValue/1e18 as eth_amount
+            , a.evt_block_time 
+            , date_trunc('week',a.evt_block_time) as evt_block_time_week
+            , a.evt_block_number
+            , a.evt_index
+            , a.evt_tx_hash
+    from {{ source('cryptopunks_ethereum','CryptoPunksMarket_evt_PunkOffered') }} a
+    inner join {{ source('ethereum','transactions') }} b on a.evt_tx_hash = b.hash
```
Comment (jeff-dude) [inline models/cryptopunks/ethereum/cryptopunks_ethereum_punk_offer_events.sql:35]: please add incremental filter within join conditions on this source

File: models/cryptopunks/ethereum/cryptopunks_ethereum_punk_offer_events.sql [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            , b.from  
+            , cast(NULL as varchar(5)) as to 
+            , cast(NULL as double) as eth_amount
+            , a.evt_block_time
+            , date_trunc('week',a.evt_block_time) as evt_block_time_week
+            , a.evt_block_number
+            , a.evt_index
+            , a.evt_tx_hash
+    from {{ source('cryptopunks_ethereum','CryptoPunksMarket_evt_PunkNoLongerForSale') }} a
+    inner join {{ source('ethereum','transactions') }} b on a.evt_tx_hash = b.hash
```
Comment (jeff-dude) [inline models/cryptopunks/ethereum/cryptopunks_ethereum_punk_offer_events.sql:53]: same filter here

**Review by jeff-dude — CHANGES_REQUESTED** [review]

**Comment by catherine-2** [comment]
Thank you @jeff-dude ! Updated for your comments

File: models/cryptopunks/ethereum/cryptopunks_ethereum_current_listings.sql (hunk 1 above)
Comment (jeff-dude) [reply models/cryptopunks/ethereum/cryptopunks_ethereum_current_listings.sql:51]: sorry i wasn't clear -- not an incremental filter, this is a view, but to make `prices.usd` a source jinja syntax. you have it hardcoded :) it helps with dependencies in order of operations

Comment (catherine-2) [reply models/cryptopunks/ethereum/cryptopunks_ethereum_current_listings.sql:51]: Ah sorry I missed that - yes will fix now!

File: models/cryptopunks/ethereum/cryptopunks_ethereum_punk_offer_events.sql (hunk 3 above)
Comment (jeff-dude) [reply models/cryptopunks/ethereum/cryptopunks_ethereum_punk_offer_events.sql:35]: i wasn't clear here either. there should be two filters:
1 - your prior where clause, that can be reverted back to table alias `a` is filtered properly
2 - an additional join condition on table alias `b` for block time incremental filter -- we need to filter ethereum tx's base table as much as possible

Comment (catherine-2) [reply models/cryptopunks/ethereum/cryptopunks_ethereum_punk_offer_events.sql]: Got it! I think I understand and fixed in the latest commits just now but let me know if it's still off!

File: models/cryptopunks/ethereum/cryptopunks_ethereum_events.sql [hunk 5]
```diff
@@ -17,11 +13,11 @@ with cryptopunks_bids_and_sales as (
     from 
     (
     select  "PunkBought" as event_type
-            , `punkIndex` as punk_id 
-            , `value`/1e18 as sale_price   
+            , punkIndex as punk_id 
+            , value/1e18 as sale_price   
             , NULL as bid_amount
```
Comment (jeff-dude) [inline models/cryptopunks/ethereum/cryptopunks_ethereum_events.sql:18]: you may want to explicitly cast all these hardcoded null values to desired datatypes, similar to other models you have

File: models/cryptopunks/ethereum/cryptopunks_ethereum_punk_offer_events.sql (hunk 3 above)
Comment (jeff-dude) [reply models/cryptopunks/ethereum/cryptopunks_ethereum_punk_offer_events.sql]: pushed a commit with a simplified approach

Comment (catherine-2) [reply models/cryptopunks/ethereum/cryptopunks_ethereum_punk_offer_events.sql]: Thank you!

File: models/cryptopunks/ethereum/cryptopunks_ethereum_events.sql (hunk 5 above)
Comment (catherine-2) [reply models/cryptopunks/ethereum/cryptopunks_ethereum_events.sql:18]: Makes sense - will update that now too!

Comment (catherine-2) [reply models/cryptopunks/ethereum/cryptopunks_ethereum_events.sql]: Updated!

**Review by jeff-dude — APPROVED** [review]

**Comment by jeff-dude** [comment]
thanks for the quick fixes/enhancements, let's get it merged 🤝

**Comment by catherine-2** [comment]
> thanks for the quick fixes/enhancements, let's get it merged 🤝

❤️🙏   Thank you so much for your help Jeff!

**Pull request merged**